### ISSUE TITLE
inconsistent  outputs for the yolo_nas_l backbone when using different batch size with the same elements 

### ISSUE THREAD

**Issue by RamShiraziAB** [Deci-AI/super-gradients#1690]
### 🐛 Describe the bug

when we use the backbone of the yolo_nas_l model with different batches (with different sizes)  with the same elements across batch, we get inconsistent results ! 
we cant use batch processing on the GPU (CPU is stable)
```python 
import numpy as np
import super_gradients
import torch

def test_yolo_nas_backbone_with_fe_batch_consistency():
    device = torch.device('cuda')
    yolo_nas = super_gradients.training.models.get("yolo_nas_l", pretrained_weights="coco")\
        .to(device)
    yolo_nas.prep_model_for_conversion(full_fusion=True)
    yolo_nas.eval()
    random_im =np.random.randint(low=0,high=255,size=(1,3,64,64),dtype=np.uint8)
    batch_input_M = torch.tensor(np.concatenate([random_im for _ in range(1)],axis=0)).to(dtype=torch.float32).to(device)
    batch_input_N = torch.tensor(np.concatenate([random_im for _ in range(2)],axis=0)).to(dtype=torch.float32).to(device)
    batch_output_per_layer_M = yolo_nas.backbone(batch_input_M) 
    batch_output_per_layer_N = yolo_nas.backbone(batch_input_N) 
    for batch_output_M,batch_output_N in zip(batch_output_per_layer_M,batch_output_per_layer_N):
        print(f"MEAN ERROR  {(torch.abs(batch_output_M[0]-batch_output_N[0])).mean()}")
        print(f"MAX ERROR  {(torch.abs(batch_output_M[0]-batch_output_N[0])).max()}")
        
if __name__ == "__main__":
    test_yolo_nas_backbone_with_fe_batch_consistency()
```

and the output for the script is :
```
MEAN ERROR  0.06328702718019485
MAX ERROR  3.005645751953125
MEAN ERROR  0.020327161997556686
MAX ERROR  1.1500320434570312
MEAN ERROR  0.004258384462445974
MAX ERROR  0.12517306208610535
MEAN ERROR  0.00048287835670635104
MAX ERROR  0.010539054870605469
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.28.0
Libc version: glibc-2.31

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA RTX A5000
Nvidia driver version: 525.147.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             24
On-line CPU(s) list:                0-23
Thread(s) per core:                 1
Core(s) per socket:                 16
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              151
Model name:                         12th Gen Intel(R) Core(TM) i9-12900
Stepping:                           2
CPU MHz:                            2400.000
CPU max MHz:                        5100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4838.40
Virtualization:                     VT-x
L1d cache:                          384 KiB
L1i cache:                          256 KiB
L2 cache:                           10 MiB
NUMA node0 CPU(s):                  0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] onnx==1.13.0
[pip3] onnx-graphsurgeon==0.3.27
[pip3] onnx-simplifier==0.4.35
[pip3] onnxruntime==1.13.1
[pip3] torch==2.1.0
[pip3] torchmetrics==0.8.0
[pip3] torchvision==0.16.0
[pip3] triton==2.1.0
[conda] numpy                     1.23.0                   pypi_0    pypi
[conda] torch                     2.1.0                    pypi_0    pypi
[conda] torchmetrics              0.8.0                    pypi_0    pypi
[conda] torchvision               0.16.0                   pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

**Comment by BloodAxe** [comment]
Need to consult with PyTorch team I assume. 
My guess is that torch may have some dispatching mechanism inside conv2d method to use different computation strategies that depends on batch, in/out channels and spatial size of the input. So when the batch size is different - a different kernels are chosen and maybe intricate nuances of their implementation cause this discrepancy. 

Please note that example input that you are using in this example has [0..255] range which is not what YoloNAS expects (Should be divided by 255). If you scale input by 255 the error drops proportionally. It is still far from zero, though.

**Comment by hinmanj** [comment]
I also just ran into this (in C++) when trying to use models that are BatchSize=1 vs. BatchSize=4 exported to ONNX.

I couldn't figure out what was going on until I switched back to the BatchSize=1 model to process all 4 images sequentially, and it worked as expected. Not really sure if there's a plan to solve, or even a solutio for this, but the reason I'm batch inferencing is to process a set of 4 images together being streamed in real-time so I'll definitely be following this to see if it can be solved.

**Comment by BloodAxe** [comment]
There is no solution to this phenomenon. Even plain pytorch execution engine (TRT and ONNXRuntime as well) use mechanisms to do choose optimal computation kernels based on data size. Those computational kernels on paper doing exact same computation (Say a + b + c) but optimized for specific batch size or spatial size. In machine arithmetics order of addion DOES matter: `a + (b+c) != (a+b)+c`. So different kernels may have varying order of operations inside hence producing (slightly) different results. 

At this point I don't see any solution to this behavior that we can apply on our side to tackle this. As been shown - ALL pytorch models as affected by this.

Feel free to check the recent paper studying this if you want to learn more:
https://openreview.net/forum?id=6zyFgr1b8Q
